storage_service/administration/validators.py

```diff
@@ -0,0 +1,44 @@
+import re
+
+from django.core.exceptions import ValidationError
+from django.utils.translation import ugettext as _
+
+
+class PasswordComplexityValidator(object):
+    """Custom password complexity validator.
+
+    To pass validation, passwords must contain at least three of:
+    - uppercase characters
+    - lowercase characters
+    - numbers
+    - special characters
+    """
+
+    HELP_TEXT = _(
+        "Your password must contain at least 3 of: uppercase characters, "
+        "lowercase characters, numbers, and special characters."
```

Comment:
I can imagine a user asking what a special character is. Do you see that as an issue here as well? Is it everything else on the keyboard? The shifted characters running 0-9 on the number pas? Is there anything possible to create on the keyboard that OWASP wouldn't like at all? https://owasp.org/www-community/password-special-characters - a small number of examples here might be enough.

Reply:
My first thought was to be prescriptive about what characters were allowed, but I kind of like the permissive definition of special characters being anything that's not a lower-case character, upper-case character, or number. It does mean we allows other characters/code points such as emojis in passwords as well, so that a valid password might be, e.g., `abcd9876🐶`. This seems acceptable to Django in my testing and by my reading of the [docs](https://docs.djangoproject.com/en/1.11/ref/contrib/auth/#django.contrib.auth.models.User.password), which say that "Raw passwords can be arbitrarily long and can contain any character". OWASP is new to me - if you think it's better that we limit to that set, I think that would be fine as well.

What I'm reading as your main thrust here - that we should give some examples - is also very reasonable I think.

Reply:
We've occasionally had access to OWASP reports on AtoM but nothing in Archivematica. It seems generally useful. My question was broadly around wondering what a special character was which you've helped me to understand. And looking in more detail at OWASP guidelines beyond that kind of glossary, they seem pretty comfortable with any and everything including white-space to that's cool: [OWASP cheat sheet - password controls](https://cheatsheetseries.owasp.org/cheatsheets/Authentication_Cheat_Sheet.html#implement-proper-password-strength-controls).

RE: examples - would be lovely! Although I'm not sure what direction you might take on that given such a broad set of characters.

Reply:
The examples are tricky. I tried a couple approaches but the help text kept getting pretty long or hard to read, so I've tried going with more of a brief explanation instead in commit https://github.com/artefactual/archivematica-storage-service/pull/553/commits/9cd4a3e9f4b88a2bf74a28fa6a45237f5007c8df:

"Your password must contain at least 3 of: uppercase characters, lowercase characters, numbers, and special characters (symbols or other Unicode characters)."

LMK if you'd rather see proper examples.

Reply:
As of commit https://github.com/artefactual/archivematica-storage-service/pull/553/commits/39599627341e4d1a3f852520e16b8dd5422c21b2 this description is now accurate, because accented alphabetical characters are treated as their ASCII equivalents.